crypto/x509/v3_purp.c

```diff
@@ -883,6 +888,58 @@ static int check_purpose_timestamp_sign(const X509_PURPOSE *xp, const X509 *x,
     return 1;
 }
 
+static int check_purpose_code_sign(const X509_PURPOSE *xp, const X509 *x,
+                                   int require_ca)
+{
+    int i_ext;
+
+    /* If ca is true we must return if this is a valid CA certificate. */
+    if (require_ca)
+        return check_ca(x);
+
+    /*
+     * Check the key usage field:
+     * Key Usage must be present, marked as critical and have the
+     * digitalSignature bit set.
+     * keyCertSign and cRLSign must not be set.
+     *
+     * Reference: CA Browser Forum
+     * Baseline Requirements for the Issuance and Management of Code Signing
+     * Appendix B: Certificate Extensions (Normative)
+     * (3) Code Signing Certificates
+     */
+    if ((x->ex_flags & EXFLAG_KUSAGE) == 0)
+        return 0;
+    if ((x->ex_kusage & KU_DIGITAL_SIGNATURE) == 0)
+        return 0;
+    if ((x->ex_kusage & (KU_KEY_CERT_SIGN | KU_CRL_SIGN)) != 0)
+        return 0;
+
+    /* Key Usage MUST be critical */
+    i_ext = X509_get_ext_by_NID(x, NID_key_usage, -1);
+    if (i_ext < 0)
+        return 0;
+    if (i_ext >= 0) {
+        X509_EXTENSION *ext = X509_get_ext((X509 *)x, i_ext);
+        if (!X509_EXTENSION_get_critical(ext))
+            return 0;
+    }
+
+    /*
+     * Extended key usage must be present and have the codeSign bit set.
+     * anyKeyUsage and serverAuth must not be present
+     */
+    if ((x->ex_flags & EXFLAG_XKUSAGE) == 0)
+        return 0;
+    if ((x->ex_xkusage & XKU_CODE_SIGN) == 0)
+        return 0;
+    if ((x->ex_xkusage & (XKU_ANYEKU | XKU_SSL_SERVER)) != 0)
+        return 0;
```

Comment:
Nice work!
Yet why do you exclude other extended key usages?
Either add a quotation from the spec that this is required or remove this check.
Usually, EKUs do not exclude others.

Reply:
I have updated the PR to now contain the exact text of the CA browser forum requirements. I did not find any other specification detailing code signing. For purposes like time stamping there are for example ETSI standards available.
Having this said, I think that it makes perfect sense to ensure that certificates have a specific purpose. Consider a mixed SSLserver and codesiging certificate that would allow the system administrator of your web site to sign code with that same certificate. As a relying party (checking the certificate purpose) it as well makes sense to ensure that no such certificate was used for signatures.

Reply:
And how about the requirements demanded by CABF on CA certs,
and the further ones on EE certs regarding policies, CDPs, and AIA?

Reply:
Well, not that I'd demand checking that very rarely used utterly complicated policy stuff,
but please make clear in the doc and comment that these checks are not performed.
Anyway, for various reasons, OpenSSL is rather relaxed/sloppy/graceful on checking cert requirements.

Reply:
I have not considered adding the requirements for the CRL distribution points and the AIA, since setting these up would be a significant hurdle for smaller organizations for example using RAUC to sign their firmware. Well, one could always check for the attributes to be present, but other tools might or will fall back and try to actually access these URLs and will then throw errors.
I also would consider it difficult to cover such things in the OpenSSL test suite. 

OpenSSL does have a history and compatibility to consider. However, for many developers OpenSSL is not just a toolbox but servs as a reference implementation, so I think we should be precise if possible.

Reply:
Sorry for warming up this comment, but I'm not sure where a better place would be.

I'm a RAUC (Linux OTA SW updater) maintainer and found this PR when our [local codesign purpose check](https://github.com/rauc/rauc/blob/cc03d35f9876ce55706c08472daaf3179a18e701/src/signature.c#L36) [started to fail with 3.2.0](https://github.com/rauc/rauc/issues/1311), as there are differences:

- OpenSSL enforces Key Usage existence, while RAUC only checks that it allows digital signatures *if it exists*
- RAUC enforces that Extended Key Usage on CA certs allows code signing if it exists

The [CAB Forum Code Singing Requirements](https://cabforum.org/wp-content/uploads/Baseline-Requirements-for-the-Issuance-and-Management-of-Code-Signing.v3.2.pdf) are even stricter, as Subordinate CA used for Code Signing MUST have it as a Extended Key Usage (7.1.2.2 f).

In the context auf RAUC, I see the CAB Forum requirement on Subordinate CAs as important: Otherwise, it's not possible to have a PKI where only specific Sub CAs can create code signing certs.

So, for now, we'd keep using our local codesign purpose check and only use a different name internally to avoid the conflict.

@ljaenicke Do you have any reason to *not* enforce the CA cert requirements? Otherwise, what check could be added and RAUC could (optionally) allow using the OpenSSL purpose check implementation in the future.

Reply:
I was reluctant to add the CAB Forum enforcement after the discussion in https://github.com/openssl/openssl/issues/18580. I was concerned that too many things might break in the field. Having this said, I would not see a reason to not create a respective PR if there is positive feedback.
Note: the work on this topic initially started when RAUC did not yet support codesign, so our work here is overlapping. And yes, our code signing certificates for RAUC do follow CA Browser Forum also including its Sub CA requirements :-).